File: android/app/src/main/java/com/claudelists/app/api/CourtListsApi.kt
package com.claudelists.app.api

import android.util.Log
import com.claudelists.app.BuildConfig
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.Serializable

private const val TAG = "CourtListsApi"

/**
 * REST API client for Court Lists backend.
 *
 * Uses Google ID token for authentication.
 * All authenticated endpoints use the token in Authorization header.
 */
class CourtListsApi(
    private val client: HttpClient,
    private val authManager: AuthManager
) {
    private val baseUrl = BuildConfig.API_BASE_URL

    private suspend fun HttpRequestBuilder.addAuth() {
        authManager.getGoogleIdToken()?.let { token ->
            header("Authorization", "Bearer $token")
        }
    }

    // =========================================================================
    // Listings & Cases (JWT required to prevent abuse)
    // =========================================================================

    /** Get court listings for a date (POST /functions/v1/listings) */
    suspend fun getListings(date: String): List<DiaryEntry> {
        val response = client.post("$baseUrl/functions/v1/listings") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(ListingsRequest(date))
        }
        return response.body()
    }

    /** Get cases for a listing (POST /functions/v1/cases) */
    suspend fun getCases(url: String): CasesResponse {
        val response = client.post("$baseUrl/functions/v1/cases") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(CasesRequest(url))
        }
        return response.body()
    }

    // =========================================================================
    // Comments
    // =========================================================================

    /** Get comments for a case (JWT required) */
    suspend fun getComments(listSourceUrl: String, caseNumber: String): List<Comment> {
        val response = client.get("$baseUrl/functions/v1/comments") {
            addAuth()
            parameter("list_source_url", listSourceUrl)
            parameter("case_number", caseNumber)
        }
        return response.body()
    }

    /** Get comment counts for multiple cases (JWT required) */
    suspend fun getCommentCounts(listSourceUrl: String, caseNumbers: List<String>): List<CommentCount> {
        val response = client.get("$baseUrl/functions/v1/comments") {
            addAuth()
            parameter("list_source_url", listSourceUrl)
            parameter("case_numbers", caseNumbers.joinToString(","))
        }
        return response.body()
    }

    @Serializable
    data class CommentCount(
        @kotlinx.serialization.SerialName("case_number") val caseNumber: String,
        val urgent: Boolean = false
    )

    /** Add a comment (auth required) */
    suspend fun addComment(listSourceUrl: String, caseNumber: String, content: String, urgent: Boolean = false) {
        client.post("$baseUrl/functions/v1/comments") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(AddCommentRequest(listSourceUrl, caseNumber, content, urgent))
        }
    }

    @Serializable
    private data class AddCommentRequest(
        @kotlinx.serialization.SerialName("list_source_url") val listSourceUrl: String,
        @kotlinx.serialization.SerialName("case_number") val caseNumber: String,
        val content: String,
        val urgent: Boolean = false
    )

    /** Delete a comment (auth required) */
    suspend fun deleteComment(id: Long) {
        client.delete("$baseUrl/functions/v1/comments") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(mapOf("id" to id))
        }
    }

    // =========================================================================
    // Case Status
    // =========================================================================

    /** Get statuses for cases (JWT required) */
    suspend fun getCaseStatuses(listSourceUrl: String, caseNumbers: List<String>): List<CaseStatus> {
        val response = client.get("$baseUrl/functions/v1/case-status") {
            addAuth()
            parameter("list_source_url", listSourceUrl)
            parameter("case_numbers", caseNumbers.joinToString(","))
        }
        return response.body()
    }

    /** Upsert case status (auth required) */
    suspend fun upsertCaseStatus(listSourceUrl: String, caseNumber: String, done: Boolean) {
        client.post("$baseUrl/functions/v1/case-status") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(UpsertStatusRequest(listSourceUrl, caseNumber, done))
        }
    }

    @Serializable
    private data class UpsertStatusRequest(
        @kotlinx.serialization.SerialName("list_source_url") val listSourceUrl: String,
        @kotlinx.serialization.SerialName("case_number") val caseNumber: String,
        val done: Boolean
    )

    // =========================================================================
    // Watched Cases
    // =========================================================================

    /** Get watched cases for current user (auth required) */
    suspend fun getWatchedCases(): List<WatchedCase> {
        val response = client.get("$baseUrl/functions/v1/watched-cases") {
            addAuth()
        }
        return response.body()
    }

    /** Watch a case (auth required) */
    suspend fun watchCase(listSourceUrl: String, caseNumber: String) {
        client.post("$baseUrl/functions/v1/watched-cases") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(WatchRequest(listSourceUrl, caseNumber, "manual"))
        }
    }

    @Serializable
    private data class WatchRequest(
        @kotlinx.serialization.SerialName("list_source_url") val listSourceUrl: String,
        @kotlinx.serialization.SerialName("case_number") val caseNumber: String,
        val source: String = "manual"
    )

    /** Unwatch a case (auth required) */
    suspend fun unwatchCase(listSourceUrl: String, caseNumber: String) {
        client.delete("$baseUrl/functions/v1/watched-cases") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(UnwatchRequest(listSourceUrl, caseNumber))
        }
    }

    @Serializable
    private data class UnwatchRequest(
        @kotlinx.serialization.SerialName("list_source_url") val listSourceUrl: String,
        @kotlinx.serialization.SerialName("case_number") val caseNumber: String
    )

    // =========================================================================
    // Notifications
    // =========================================================================

    /** Get notifications for current user (auth required) */
    suspend fun getNotifications(limit: Int = 50): List<AppNotification> {
        val response = client.get("$baseUrl/functions/v1/notifications") {
            addAuth()
            parameter("limit", limit.toString())
        }
        return response.body()
    }

    /** Mark notification as read (auth required) */
    suspend fun markNotificationRead(id: Long) {
        client.patch("$baseUrl/functions/v1/notifications") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(mapOf("id" to id))
        }
    }

    /** Mark all notifications as read (auth required) */
    suspend fun markAllNotificationsRead() {
        client.patch("$baseUrl/functions/v1/notifications") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(mapOf("all" to true))
        }
    }

    /** Delete a notification (auth required) */
    suspend fun deleteNotification(id: Long) {
        client.delete("$baseUrl/functions/v1/notifications") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(mapOf("id" to id))
        }
    }

    /** Delete all notifications (auth required) */
    suspend fun deleteAllNotifications() {
        client.delete("$baseUrl/functions/v1/notifications") {
            addAuth()
            contentType(ContentType.Application.Json)
            setBody(mapOf("all" to true))
        }
    }
}

File: ios/CourtLists/CourtLists/Services/CourtListsAPI.swift
import Foundation

enum APIError: Error {
    case invalidURL
    case requestFailed(Error)
    case decodingFailed(Error)
    case httpError(Int)
    case unauthorized
}

class CourtListsAPI {
    private let baseUrl: String
    private let session: URLSession
    private weak var authManager: AuthManager?

    init(baseUrl: String = Configuration.apiBaseUrl, authManager: AuthManager? = nil) {
        self.baseUrl = baseUrl
        self.session = URLSession.shared
        self.authManager = authManager
    }

    func setAuthManager(_ authManager: AuthManager) {
        self.authManager = authManager
    }

    // MARK: - Helper Methods

    private func createRequest(url: URL, method: String = "GET") async -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        // Always add auth token - all endpoints require JWT
        if let authManager = authManager {
            await authManager.refreshTokenIfNeeded()
            if let token = await MainActor.run(body: { authManager.idToken }) {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }
        }

        return request
    }

    private func performRequest<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.requestFailed(NSError(domain: "Invalid response", code: -1))
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 {
                throw APIError.unauthorized
            }
            throw APIError.httpError(httpResponse.statusCode)
        }

        do {
            let decoder = JSONDecoder()
            return try decoder.decode(T.self, from: data)
        } catch {
            if let rawString = String(data: data, encoding: .utf8) {
                print("Decoding failed. Raw response: \(rawString.prefix(500))")
            }
            print("Decoding error: \(error)")
            throw APIError.decodingFailed(error)
        }
    }

    private func performRequestNoResponse(_ request: URLRequest) async throws {
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.requestFailed(NSError(domain: "Invalid response", code: -1))
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if httpResponse.statusCode == 401 {
                throw APIError.unauthorized
            }
            if let rawString = String(data: data, encoding: .utf8) {
                print("Request failed. Response: \(rawString.prefix(500))")
            }
            throw APIError.httpError(httpResponse.statusCode)
        }
    }

    // MARK: - Listings & Cases (Public)

    func getListings(date: String) async throws -> [DiaryEntry] {
        guard let url = URL(string: "\(baseUrl)/functions/v1/listings") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "POST")
        request.httpBody = try JSONEncoder().encode(ListingsRequest(date: date))

        return try await performRequest(request)
    }

    func getCases(url: String) async throws -> CasesResponse {
        guard let apiUrl = URL(string: "\(baseUrl)/functions/v1/cases") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: apiUrl, method: "POST")
        request.httpBody = try JSONEncoder().encode(CasesRequest(url: url))

        return try await performRequest(request)
    }

    // MARK: - Comments

    func getComments(listSourceUrl: String, caseNumber: String) async throws -> [Comment] {
        guard var components = URLComponents(string: "\(baseUrl)/functions/v1/comments") else {
            throw APIError.invalidURL
        }

        components.queryItems = [
            URLQueryItem(name: "list_source_url", value: listSourceUrl),
            URLQueryItem(name: "case_number", value: caseNumber)
        ]

        guard let url = components.url else {
            throw APIError.invalidURL
        }

        let request = await createRequest(url: url)
        return try await performRequest(request)
    }

    func getCommentCounts(listSourceUrl: String, caseNumbers: [String]) async throws -> [CommentCount] {
        guard var components = URLComponents(string: "\(baseUrl)/functions/v1/comments") else {
            throw APIError.invalidURL
        }

        components.queryItems = [
            URLQueryItem(name: "list_source_url", value: listSourceUrl),
            URLQueryItem(name: "case_numbers", value: caseNumbers.joined(separator: ","))
        ]

        guard let url = components.url else {
            throw APIError.invalidURL
        }

        let request = await createRequest(url: url)
        return try await performRequest(request)
    }

    func addComment(listSourceUrl: String, caseNumber: String, content: String, urgent: Bool = false) async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/comments") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "POST")

        let body: [String: Any] = [
            "list_source_url": listSourceUrl,
            "case_number": caseNumber,
            "content": content,
            "urgent": urgent
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        try await performRequestNoResponse(request)
    }

    func deleteComment(id: Int64) async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/comments") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "DELETE")
        request.httpBody = try JSONEncoder().encode(["id": id])

        try await performRequestNoResponse(request)
    }

    // MARK: - Case Status

    func getCaseStatuses(listSourceUrl: String, caseNumbers: [String]) async throws -> [CaseStatus] {
        guard var components = URLComponents(string: "\(baseUrl)/functions/v1/case-status") else {
            throw APIError.invalidURL
        }

        components.queryItems = [
            URLQueryItem(name: "list_source_url", value: listSourceUrl),
            URLQueryItem(name: "case_numbers", value: caseNumbers.joined(separator: ","))
        ]

        guard let url = components.url else {
            throw APIError.invalidURL
        }

        let request = await createRequest(url: url)
        return try await performRequest(request)
    }

    func upsertCaseStatus(listSourceUrl: String, caseNumber: String, done: Bool) async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/case-status") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "POST")

        let body: [String: Any] = [
            "list_source_url": listSourceUrl,
            "case_number": caseNumber,
            "done": done
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        try await performRequestNoResponse(request)
    }

    // MARK: - Watched Cases

    func getWatchedCases() async throws -> [WatchedCase] {
        guard let url = URL(string: "\(baseUrl)/functions/v1/watched-cases") else {
            throw APIError.invalidURL
        }

        let request = await createRequest(url: url)
        return try await performRequest(request)
    }

    func watchCase(listSourceUrl: String, caseNumber: String) async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/watched-cases") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "POST")

        let body: [String: String] = [
            "list_source_url": listSourceUrl,
            "case_number": caseNumber,
            "source": "manual"
        ]
        request.httpBody = try JSONEncoder().encode(body)

        try await performRequestNoResponse(request)
    }

    func unwatchCase(listSourceUrl: String, caseNumber: String) async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/watched-cases") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "DELETE")

        let body: [String: String] = [
            "list_source_url": listSourceUrl,
            "case_number": caseNumber
        ]
        request.httpBody = try JSONEncoder().encode(body)

        try await performRequestNoResponse(request)
    }

    // MARK: - Notifications

    func getNotifications(limit: Int = 50) async throws -> [AppNotification] {
        guard var components = URLComponents(string: "\(baseUrl)/functions/v1/notifications") else {
            throw APIError.invalidURL
        }

        components.queryItems = [
            URLQueryItem(name: "limit", value: "\(limit)")
        ]

        guard let url = components.url else {
            throw APIError.invalidURL
        }

        let request = await createRequest(url: url)
        return try await performRequest(request)
    }

    func markNotificationRead(id: Int64) async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/notifications") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "PATCH")
        request.httpBody = try JSONEncoder().encode(["id": id])

        try await performRequestNoResponse(request)
    }

    func markAllNotificationsRead() async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/notifications") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "PATCH")
        request.httpBody = try JSONEncoder().encode(["all": true])

        try await performRequestNoResponse(request)
    }

    func deleteNotification(id: Int64) async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/notifications") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "DELETE")
        request.httpBody = try JSONEncoder().encode(["id": id])

        try await performRequestNoResponse(request)
    }

    func deleteAllNotifications() async throws {
        guard let url = URL(string: "\(baseUrl)/functions/v1/notifications") else {
            throw APIError.invalidURL
        }

        var request = await createRequest(url: url, method: "DELETE")
        request.httpBody = try JSONEncoder().encode(["all": true])

        try await performRequestNoResponse(request)
    }
}
